File: GFG/BST/fixbst.cpp
#include<iostream>
#include"../ds.h"
using namespace std;

TNode<int>* pre=NULL, *first=NULL, *second=NULL;
void fixb(TNode<int>* root){
    if(root==NULL){
        return;
    }
    fixb(root->left);
    if(pre!=NULL && root->key<pre->key){
        if(first==NULL){
            first=pre;
        }
        second=root;
    }
    pre=root;
    fixb(root->right);
}
void inorder(TNode<int>* root){
    if(root==NULL){
        return ;
    }
    inorder(root->left);
    cout<<root->key<<" ";
    inorder(root->right);
}

int main(){
    TNode<int> * root= new TNode(20);
    root->left=new TNode(60);
    root->left->left=new TNode(4);
    root->left->right=new TNode(10);
    root->right=new TNode(80);
    root->right->left=new TNode(8);
    root->right->right=new TNode(100);
    inorder(root);
    fixb(root);
    swap(first->key,second->key);
    cout<<endl;
    inorder(root);
    return 0;
}

File: GFG/BST/fixbst.py
import ds

def fixb(root):
    global pre,first,second
    if(root is None):
        return
    fixb(root.left)
    if(pre!=None and root.key<pre.key):
        if(first is None):
            first=pre
        second=root
    pre=root
    fixb(root.right)
    
def inorder(root):
    if(root is None):
        return
    inorder(root.left)
    print(root.key,end=" ")
    inorder(root.right)    

    
root= ds.TNode(20)
root.left=ds.TNode(60)
root.left.left=ds.TNode(4)
root.left.right=ds.TNode(10)
root.right=ds.TNode(80)
root.right.left=ds.TNode(8)
root.right.right=ds.TNode(100)
pre=root.right.right.right
first=root.right.right.right
second=root.right.right.right
inorder(root)
fixb(root)
print()
first.key,second.key=second.key,first.key
inorder(root)